libmesabox/src/lsb/mount/mod.rs

```diff
@@ -0,0 +1,803 @@
+                dir.to_string_lossy().to_string(),
+            ))?.as_path())
+    }
+}
+
+lazy_static! {
+    static ref OPTION_MAP: HashMap<OsString, c_ulong> = {
+        let mut option_map = HashMap::new();
+        option_map.insert(OsString::from("async"), !libc::MS_SYNCHRONOUS);
+        option_map.insert(OsString::from("atime"), !libc::MS_NOATIME);
+        option_map.insert(OsString::from("dev"), !libc::MS_NODEV);
+        option_map.insert(OsString::from("exec"), !libc::MS_NOEXEC);
+        option_map.insert(OsString::from("noatime"), libc::MS_NOATIME);
+        option_map.insert(OsString::from("nodev"), libc::MS_NODEV);
+        option_map.insert(OsString::from("noexec"), libc::MS_NOEXEC);
+        option_map.insert(OsString::from("nosuid"), libc::MS_NOSUID);
+        option_map.insert(OsString::from("remount"), libc::MS_REMOUNT);
+        option_map.insert(OsString::from("ro"), libc::MS_RDONLY);
+        option_map.insert(OsString::from("rw"), !libc::MS_RDONLY);
+        option_map.insert(OsString::from("suid"), !libc::MS_NOSUID);
+        option_map.insert(OsString::from("sync"), libc::MS_SYNCHRONOUS);
+        option_map
+    };
+}
+
+struct Flag {
+    flag: MsFlags,
+}
+
+impl Default for Flag {
+    fn default() -> Self {
+        Self {
+            flag: MsFlags::MS_SILENT,
+        }
+    }
+}
+
+impl Flag {
+    fn from(options: &mut Vec<OsString>) -> MountResult<MsFlags> {
+        //let mut flag = Flag::default();
+        let mut flag = Flag::default().flag.bits();
+        if options.contains(&OsString::from("default")) {
+            options.extend_from_slice(&[
+                OsString::from("rw"),
+                OsString::from("suid"),
+                OsString::from("dev"),
+                OsString::from("exec"),
+                OsString::from("auto"),
+                OsString::from("nouser"),
+                OsString::from("async"),
+            ]);
+        }
+        for opt in options {
+            let f = *OPTION_MAP.get(opt).ok_or(MountError::UnsupportedOption)?;
+            if (f as c_long) < 0 {
+                flag &= f;
+            } else {
+                flag |= f;
+            }
+        }
+        Ok(MsFlags::from_bits(flag).ok_or(MountError::UnsupportedOption)?)
+    }
+}
+
+struct MntEnt {
+    mnt_fsname: OsString,
+    mnt_dir: OsString,
+    mnt_type: OsString,
+    mnt_opts: OsString,
+}
+
+/// This is used to read Filesystem Description File
+/// e.g. /etc/fstab and /etc/mtab
+struct FSDescFile {
+    entries: Vec<MntEnt>,
+}
+
+impl FSDescFile {
+    fn new(path: &str) -> MountResult<Self> {
+        // all of these files should exist and can be read, but just in case
+        let file = fs::File::open(path).or(Err(MountError::OpenFileError(String::from(path))))?;
+        let mut entries = Vec::new();
+        for line in BufReader::new(file).lines() {
+            let line = line.or(Err(MountError::FSDescFileFormatError(String::from(path))))?;
```

Comment:
I don't think these files are guaranteed to be valid UTF-8, so I think using `String`s here is wrong.

Reply:
You are right. If these files contain invalid UTF-8, the program will return an error. Now I use `read_until` to read them in bytes.